packages/typir/src/kinds/class-kind.ts

```diff
@@ -550,3 +604,232 @@ export class UniqueClassValidation implements ValidationRuleWithBeforeAfter {
+        isMethodDeclaration: (domainElement: unknown) => domainElement is T,
+        getClassOfMethod: (domainElement: T, methodType: FunctionType) => unknown) {
+        this.services = services;
+        this.isMethodDeclaration = isMethodDeclaration;
+        this.getClassOfMethod = getClassOfMethod;
+    }
+
+    beforeValidation(_domainRoot: unknown, _typir: TypirServices): ValidationProblem[] {
+        this.foundDeclarations.clear();
+        return [];
+    }
+
+    validation(domainElement: unknown, _typir: TypirServices): ValidationProblem[] {
+        if (this.isMethodDeclaration(domainElement)) { // improves performance, since type inference need to be done only for relevant elements
+            const methodType = this.services.inference.inferType(domainElement);
+            if (isFunctionType(methodType)) {
+                const classDeclaration = this.getClassOfMethod(domainElement, methodType);
+                const classType = this.services.inference.inferType(classDeclaration);
+                if (isClassType(classType)) {
+                    const key = this.calculateMethodKey(classType, methodType);
+                    let entries = this.foundDeclarations.get(key);
+                    if (!entries) {
+                        entries = [];
+                        this.foundDeclarations.set(key, entries);
+                    }
+                    entries.push(domainElement);
+                }
+            }
+        }
+        return [];
+    }
+
+    /**
+     * Calculates a key for a method which encodes its unique properties, i.e. duplicate methods have the same key.
+     * Additionally, the class of the method needs to be represented in the key as well.
+     * This key is used to identify duplicated methods.
+     * Override this method to change the properties which make a method unique.
+     * @param clas the current class type
+     * @param func the current function type
+     * @returns a string key
+     */
+    protected calculateMethodKey(clas: ClassType, func: FunctionType): string {
+        return `${clas.identifier}.${func.functionName}(${func.getInputs().map(param => param.type.identifier)})`;
+    }
+
+    afterValidation(_domainRoot: unknown, _typir: TypirServices): ValidationProblem[] {
+        const result: ValidationProblem[] = [];
+        for (const [key, methods] of this.foundDeclarations.entries()) {
+            if (methods.length >= 2) {
+                for (const method of methods) {
+                    result.push({
+                        $problem: ValidationProblem,
+                        domainElement: method,
+                        severity: 'error',
+                        message: `Declared methods need to be unique (${key}).`,
+                    });
+                }
+            }
+        }
+
+        this.foundDeclarations.clear();
+        return result;
+    }
+}
+
+
+// TODO for the review: which name is better? AnyClassType vs TopClassType?
```

Comment:
I have no strong personal preference. TopClassType leaves no doubt what this is for but AnyClassType is nicer regarding natural language imho. However, maybe we should just decide based on consistency.

Reply:
Thanks for your argumentation! I agree: To be consistent with `TopType`, I renamed `AnyClassType` to `TopClassType`.